README.md
# Sologenic NFT SDK

This library is to (bulk) mint NFTs on the Sologenic NFT marketplace (sologenic.org/nfts/marketplace).

After the minting process is completed, the NFTs will take a few minutes show up in the marketplace.

DUE TO THE REQUIREMENTS OF USING THE WALLET SECRET, DO NOT USE THIS LIBRARY AT THE CLIENT SIDE !!!!!!!!!!

```
npm i sologenic-nft-sdk
```

## Contents

- [Usage](#usage)
- [Methods](#methods)
  - [getApiURL](#getApiURL)
  - [setAccount](#setAccount)
  - [getAccountNFTS](#getAccountNfts)
  - [getNFTData](#getNFTData)
  - [getWalletAddress](#getWalletAddress)
  - [getBurnConfiguration](#getBurnConfiguration)
  - [generateNFTSlots](#generateNFTSlots)
  - [createCollection](#createCollection)
  - [updateCollection](#updateCollection)
  - [getAllCollections](#getAllCollections)
  - [getCollectionAddress](#getCollectionAddress)
  - [getCollectionData](#getCollectionData)
  - [getCollectionNFTSlots](#getCollectionNFTSlots)
  - [mint](#mint)
  - [mintMultipleCopies](#mintmultiplecopies)

## Usage

```js
import { SologenicNFTManager } from "sologenic-nft-sdk";
import fs from "fs";

const minter = new SologenicNFTManager({
  mode: "mainnet" | "testnet" | "devnet",
  xrpl_node: XRPL_NODE,
});

// Example of how to submit the files
const collectionCoverBuffer = fs.readFileSync("PATH_TO_COLLECTION_COVER_FILE");
const collectionThumbnailBuffer = fs.readFileSync("PATH_TO_COLLECTION_THUMBNAIL_FILE");
const nftFileBuffer = fs.readFileSync("PATH_TO_NFT_FILE");
const nftThumbnailBuffer = fs.readFileSync("PATH_TO_NFT_THUMBNAIL_FILE");

// Set the account to mint with
const connected_account = minter.setAccount(YOUR_WALLET_SECRET);

// After initializing the Minter, we need to set the collection address in which we want to mint
// Use setCollectionAddress() if you know the collection you want to mint in
// minter.setCollectionAddress(YOUR_DESIRED_COLLECTION);

// or create a new collection with createCollection(). This method sets the recently created collection as the default to mint on, it can be overriden with setCollectionAddress().
await minter.createCollection({
  name: "Test Collection", // REQUIRED: The name of the collection
  description: "Collection description", // OPTIONAL: The description of the collection
  cover: collectionCoverBuffer, // Cover of the Collection, can be only JPG, JPEG, PNG or GIF, recommended to be a compressed version, as this is for display purposes, it's not stored on the ledger
  thumbnail: collectionThumbnailBuffer // Thumbnail of the Collection, can be only JPG, JPEG, PNG or GIF, recommended to be a compressed version, as this is for display purposes, it's not stored on the ledger
  transfer_fee: 10000, // OPTIONAL (Defaults to 0): Sets the Royalty percentage for all the NFTs within this collection. Can be overwritten per NFT. A number between 0 and 50000 i.e to get 10% royalty transfer_fee  must be 10000
});

// Once the collection is set. If you have available Slot (or burns) you can start minting.
const { mint_tx_hash, NFTokenID } = await minter.mint({
  file: nftFileBuffer, // Original data of the NFT can be any of the supported files Sologenic accepts
  thumbnail: nftThumbnailBuffer, // Thumbnail of the NFT, can be only JPG, JPEG, PNG or GIF, recommended to be a compressed version, as this is for display purposes, it's not stored on the ledger
  name: "Testing NFT 3", // REQUIRED: The name of the NFT
  category: "arts", // REQUIRED: Sets the category on the NFT for the Marketplace
  only_xrp: false, // REQUIRED (Defaults to false): Sets if the NFT can only be traded for XRP
  is_explicit: false, // REQUIRED (Default to false): Sets if the NFT contains Explicit content
  transfer_fee: 10000, // OPTIONAL (Defaults to 0): A number between 0 and 50000 i.e to get 10% royalty transfer_fee must be 10000. If set, it will override the Collection Transfer fee.
  description: "Testing NFT description", // OPTIONAL: Description of the NFT
  external_url: "https://sologenic.org", // OPTIONAL: URL for more information of the NFT
  attributes: [
    //OPTIONAL: Attributes of the NFT, what makes it unique
    {
      trait_type: "attribute",
      value: "attr",
    },
    {
      trait_type: "attribute 2",
      value: 1,
      max_value: 2,
    },
  ],
});
```

## Methods

### `getApiURL`

Returns information about the Sologenic API mode connection

```js
const url = minter.getApiURL();
```

_Response_

This method returns an object with the following properties:

| Property |                        Value |
| :------- | ---------------------------: |
| mode     | mainnet or testnet or devnet |
| url      |            Sologenic Api URL |

### `setAccount`

This method sets the default account to use.

```js
const wallet_address = minter.setAccount(YOUR_WALLET_SECRET);
// rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG
```

### `getAccountNFTS`

This method returns ALL the NFTs owned by this account in reference to the ledger; to get the NFT metadata (name, description, etc) you will need to call the method [getNFTData](#getNFTData).

```js
const nfts = await minter.getAccountNFTS();
```

_Params_

This method takes one OPTIONAL parameter; an account address. If they account address is not passed, it will default to the connected account. If there is no parameter passed nor account connected (See [setAccount](#setAccount)), this method will throw an exception.

| Param   |   Type |
| :------ | -----: |
| address | string |

_Response_
This method returns an array of the NFTs owned by the passed account (or default account) on reference with the XRP Ledger. These are the properties on each NFT object

| Property       | Description                                                                                                                                                                                                                                                                                                   |                                                                                                                                                                                      Example |
| :------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| TransferFee    | The royalty set for this NFT. 10000 = 10%                                                                                                                                                                                                                                                                     |                                                                                                                                                                                        10000 |
| Flags          | Flags are properties or other options associated with the NFT object.                                                                                                                                                                                                                                         |                                                                                                                                                                                            8 |
| NFTokenTaxon   | The taxon associated with the token.                                                                                                                                                                                                                                                                          |                                                                                                                                                                                        34356 |
| NFTokenID      | Unique identifier of the NFT on the XRPL Ledger                                                                                                                                                                                                                                                               |                                                                                                                             000A2134C4E16036D649C037D2DE7C58780DE1D985EEB986483AE3C9000001AC |
| nft_serial     | Serial number of the NFT                                                                                                                                                                                                                                                                                      |                                                                                                                                                                                            8 |
| Issuer         | Minter of the NFT.                                                                                                                                                                                                                                                                                            |                                                                                                                                                           rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG |
| URI (Optional) | Up to 256 bytes of arbitrary data. You can use the xrpl.convertHexToString utility to convert the HEX string to its string equivalent. The contents could decode to an HTTP or HTTPS URL, an IPFS URI, a magnet link, immediate data encoded as an RFC 2379 "data" URL , or even an issuer-specific encoding. | 68747470733A2F2F697066732E696F2F697066732F62616679626569666F76696766743578636B6A696C6C6A666E636F756F74716A626E7669697070326F636A35323471347675747161736D6B7378612F6D657461646174612E6A736F6E |

### `getNFTData`

This method returns the data of an specific NFT using its NFTokenID.
Note: If the NFT exists on the XRP Ledger but this method returns a NFT_NOT_FOUND, maybe the NFT was not minted on the Sologenic NFT Marketplace.

```js
const nft_data = await minter.getNFTData(NFTokenID);
```

_Params_

This method takes one parameter.

| Property  |   Type | Description                                                      |
| :-------- | -----: | :--------------------------------------------------------------- |
| NFTokenID | string | 000A2134C4E16036D649C037D2DE7C58780DE1D985EEB986483AE3C9000001AC |

_Response_

This method returns an object with 2 objects inside. `sologenic_info` and `xrpl_info`. If the NFT exists, `xrpl_info` should always be returned, if `sologenic_info` is `NULL` then it means the NFT was not minted on the sologenic platform.

#### `xrpl_info`

| Property     |                                                                                           Description | Example                                                                                        |
| :----------- | ----------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------- |
| nft_id       |                                                                                             NFTokenID | 000827107022610A05BAA45AE04D5B022D1FF298795EF9AB000083F100000000                               |
| ledger_index |                                                               Current Ledger index of the XRPL Server | 76545315                                                                                       |
| is_burned    |                                                                Whether the NFT has been burned or not | `false`                                                                                        |
| issuer       |                                                                                       Issuing address | rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG                                                             |
| owner        |                                                                                 Current owner address | rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG                                                             |
| flags        |                                                              Flags of the settings enabled in the NFT | 8                                                                                              |
| transfer_fee |                                        Royalty percentage the issuer gets after every sale of the NFT | 10000 (10%)                                                                                    |
| nft_taxon    |                                                  Taxon for the NFT set on the NFTokenMint transaction | 12345                                                                                          |
| nft_sequence |                                            Serial number of the NFT. Referenced to the Issuer account | 12                                                                                             |
| uri          |                                   Arbitrary data set on the URI field of the NFTokenMint transaction. | https://ipfs.io/ipfs/bafybeie7cd4s2pv4e5tode7xjvqjlkpp5kuflqcl73keadixevqjoq3r3y/metadata.json |
| validated    | States that the ledger index when this request was made has been validated on the XRP Ledger servers. | true                                                                                           |

#### `sologenic_info`

| Property               |                                                          Description | Example                                                                               |
| :--------------------- | -------------------------------------------------------------------: | :------------------------------------------------------------------------------------ |
| id                     |                                                            NFTokenID | 000827107022610A05BAA45AE04D5B022D1FF298795EF9AB000083F100000000                      |
| standard               |                                               NFT Standard minted on | XLS-20d                                                                               |
| collection_id          |                                                   Collection address | rDZj8PN21gmtWwqFY1SAqyubRHeT87NSVC                                                    |
| minter                 |                                                       Minter address | rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG                                                    |
| owner                  |                                                Current owner address | rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG                                                    |
| ipfs_cid               |                                               CID of the NFT on IPFS | bafybeihly7rjgfypze7hcruuss3vgzl7wh2e7zcplsbsj75ecrnuwjlydi                           |
| md5_hash               |                                  MD5 hash of the contents of the NFT | cfe2892441f41d6ea15f1e4e71614f3f                                                      |
| minted_txid            | Transaction Hash of the NFTokenMint Transaction that minted this NFT | 590B998100CCA720D0BA77BC73EF69E6358A53AE9A86A8E706DAC8498DBA0738                      |
| metadata.animation_url |                                        IPFS Url of the NFT Animation | ipfs://ipfs/bafybeiacwbv46qb7mnnjppxpshenoo5dc65uvucz3nggfimriz2hoqtsiq/animation.png |
| metadata.attributes    |                               Array of attributes pertaining the NFT | See [mint()](#mint) for details of this property                                      |
| metadata.category      |                                                  Category of the NFT | art                                                                                   |
| metadata.content_type  |                                                 Mime type of the NFT | image/png                                                                             |
| metadata.description   |                                               Description of the NFT | Testing NFT Description                                                               |
| metadata.external_url  |                                       External URL minted on the NFT | https://sologenic.org                                                                 |
| metadata.image_url     |                                            IPFS URL of the NFT Image | ipfs://ipfs/bafybeiacwbv46qb7mnnjppxpshenoo5dc65uvucz3nggfimriz2hoqtsiq/image.png     |
| metadata.is_explicit   |               Whether the NFT includes explicit or sensitive content | `false`                                                                               |
| metadata.md5hash       |                                  MD5 hash of the contents of the NFT | cfe2892441f41d6ea15f1e4e71614f3f                                                      |
| metadata.name          |                                                      Name of the NFT | Testing NFT                                                                           |

### `getWalletAddress`

Returns the address of the wallet which was used to initialize the Sologenic Minter

```js
const address = minter.getWalletAddress();
// rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG
```

### `getBurnConfiguration`

Returns the current burn configuration to get NFT Slots

```js
const config = await minter.getBurnConfiguration();
```

Response

| Property      |                                    Value |
| :------------ | ---------------------------------------: |
| burn_amount   |                                        3 |
| burn_currency | 534F4C4F00000000000000000000000000000000 |
| burn_issuer   |       rsoLo2S1kiGeCcn6hCUXVrCpGMWLrRrLZz |

### `generateNFTSlots`

Generates NFT Slots by burning SOLO.
NOTE: This method requires the connected wallet to have a positive SOLO Balance

```js
const slots_generation = await generateNFTSlots(1);
```

_Params_

| Param  | Type | Required | Example |
| :----- | :--- | :------: | :-----: |
| amount | int  |  `true`  |    1    |

_Response_

| Property    |                 Description |                                                          Example |
| :---------- | --------------------------: | ---------------------------------------------------------------: |
| address     | Wallet address used to burn |                               rBDu1BC6f1SKRvRxPiHZdeML5CRwByQTFG |
| hash        |       Burn Transaction Hash | 1016EC6C70C9C2CD9D3C6B8A8439AA538409181DBE1E5D56698BDAE61AE03608 |
| burns_count |    Amount of Slots acquired |                                                                1 |

### `createCollection`

Creates a new collection or returns the current unfinished collection.
NOTE: If there is an unfinished collection, the data passed will override the current collection information.
NOTE: This method sets the recently created (or updated) collection as the default to mint NFTs on

```js
import fs from "fs";

// Example of how to submit the files
const collectionCoverBuffer = fs.readFileSync("PATH_TO_COLLECTION_COVER_FILE");
const collectionThumbnailBuffer = fs.readFileSync("PATH_TO_COLLECTION_THUMBNAIL_FILE");


const collectionData = {
  name: "New Collection",
  description: "New Collection Description"
  cover: collectionCoverBuffer,
  thumbnail: collectionThumbnailBuffer,
  transfer_fee: 10000 // 10%
};

const new_collection = await minter.createCollection(collectionData);
```

_Params_

This method takes an object with all the data for the Collection.

| Property     | Type   | Required | Description                                                                                    |                  Example                   |
| :----------- | :----- | :------: | :--------------------------------------------------------------------------------------------- | :----------------------------------------: |
| name         | string |  `true`  | The name for the collection.                                                                   |          "My Awesome Collection"           |
| description  | string | `false`  | The description for the collection.                                                            | "My Awesome Description for my Collection" |
| cover        | Buffer | `false`  | Buffer for the cover of the collection                                                         |       <Buffer 08 06 07 05 03 00 09>        |
| thumbnail    | Buffer | `false`  | Buffer for the thumbnail of the collection                                                     |       <Buffer 08 06 07 05 03 00 09>        |
| transfer_fee | int    | `false`  | Royalties for the NFTs within this collection. NOTE: This can be overriden when minting an NFT |                   10000                    |

_Response_

| Property      |                                                                                                  Description |                                                                                   Example |
| :------------ | -----------------------------------------------------------------------------------------------------------: | ----------------------------------------------------------------------------------------: |
| issuer        |                                                                                 Identifier of the collection |                                                        rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg |
| name          |                                                                                       Name of the collection |                                                                   "My Awesome collection" |
| description   |                                                                                Description of the collection |                                                "My Awesome Description for my collection" |
| nfts          | List of NFT Slots, any slot can be empty (available) or filled (an nft has been uploaded but not minted yet) |                                     See [`getCollectionNFTSlots`](#getcollectionnftslots) |
| cover         |                                                                                    URL of the uploaded cover |     https://storage.googleapis.com/sg-nft-images/cover/rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg |
| thumbnail     |                                                                                URL of the uploaded thumbnail | https://storage.googleapis.com/sg-nft-images/thumbnail/rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg |
| unusued_burns |                                                                            Amount of available slots to mint |                                                                                         5 |
| minted_nfts   |                                                              List of ALL minted NFTs within this collections |                                                                                     NFT[] |

### `updateCollection`

Updates the current set collection.

```js
import fs from "fs";

// Example of how to submit the files
const collectionCoverBuffer = fs.readFileSync("PATH_TO_COLLECTION_COVER_FILE");
const collectionThumbnailBuffer = fs.readFileSync("PATH_TO_COLLECTION_THUMBNAIL_FILE");


const collectionData = {
  name: "New Updated Collection",
  description: "New Collection Description Update"
  cover: collectionCoverBuffer,
  thumbnail: collectionThumbnailBuffer,
  transfer_fee: 10000 // 10%
};

// Set the default collection
await minter.setCollectionAddress("rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg");

// Update default collection
await minter.updateCollection(collectionData);

// Get updated default collection
const collection = await minter.getCollectionData();

```

_Params_

This method takes an object with all the data for the Collection.
NOTE: Only unfinished collections can be updated.

| Property     | Type   | Required | Description                                                                                    |                  Example                   |
| :----------- | :----- | :------: | :--------------------------------------------------------------------------------------------- | :----------------------------------------: |
| name         | string |  `true`  | The name for the collection.                                                                   |          "My Awesome Collection"           |
| description  | string | `false`  | The description for the collection.                                                            | "My Awesome Description for my Collection" |
| cover        | Buffer | `false`  | Buffer for the cover of the collection                                                         |       <Buffer 08 06 07 05 03 00 09>        |
| thumbnail    | Buffer | `false`  | Buffer for the thumbnail of the collection                                                     |       <Buffer 08 06 07 05 03 00 09>        |
| transfer_fee | int    | `false`  | Royalties for the NFTs within this collection. NOTE: This can be overriden when minting an NFT |                   10000                    |

_Response_

This method doesn't return a response. After calling this method, you need to call `await minter.getCollectionData()` to see the updated collection data.

### `getAllCollections`

This method returns an Array of all the collections available to the account.

```js
const collections = await minter.getAllCollections();
```

_Response_

An array of Collection Objects. These are the properties of the Collection object. Please see [`getCollectionData`](#getcollectiondata) to see properties of the Collection object

### `getCollectionAddress`

This method returns the address of the default collection

```js
const collection_address = await minter.getCollectionAddress();
// rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg
```

### `getCollectionData`

This method returns the data of the default collection

_Response_

| Property     | Description                                                                                      |                                                                                     Example |
| :----------- | :----------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------: |
| name         | Name of the collection                                                                           |                                                                     "My Awesome Collection" |
| description  | Description of the collection                                                                    |                            1016EC6C70C9C2CD9D3C6B8A8439AA538409181DBE1E5D56698BDAE61AE03608 |
| cover        | URL of the cover uploaded                                                                        |     "https://storage.googleapis.com/sg-nft-images/cover/rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg" |
| thumbnail    | URL of the thumbnail uploaded                                                                    | "https://storage.googleapis.com/sg-nft-images/thumbnail/rN7DsVCsJ6vRQtNMD8yMLeKvovSXM5P3Dg" |
| transfer_fee | Royalty set for all the NFTs within this collection. NOTE: Can be overriden when minting an NFT. |                                                                                       10000 |
| cid          | Collection ID on IPFS                                                                            |                                 bafybeiba6su6ni3dfwhmjbzi5fz4pqehtnnwgg5yjwhss6mak3wqnwglhm |
| created_at   | Date of creation                                                                                 |                                                                    2022-12-30T17:23:26.000Z |
| minted_nfts  | Amount of NFTs minted in this collection                                                         |                                                                                           5 |

### `getCollectionNFTSlots`

This method returns an Array of all the NFT Slots of the default collection, both available and already uploaded, but not minted, NFTs.

_Response_

These are the properties of each NFT Slot.

| Property     |                                                                                                                                                  Description |                                                                                         Example |
| :----------- | -----------------------------------------------------------------------------------------------------------------------------------------------------------: | ----------------------------------------------------------------------------------------------: |
| id           |                                                                                                                                               ID of the slot |                                                                                           12345 |
| uid          |                                                                                                                                              UID of the slot |                                                            829df87d-0e2e-48f6-8387-b6a2e1a26462 |
| currency     |                                                                                                                                        Legacy ID for the NFT |                                                        023031516D656A664E4654393331363900000000 |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.9](https://github.com/sologenic/sologenic-nft-sdk/compare/v1.0.1...v1.0.9) (2023-01-05)

### [1.0.8](https://github.com/sologenic/sologenic-nft-sdk/compare/v1.0.1...v1.0.8) (2023-01-05)

### [1.0.4](https://github.com/AlOaks/minter-agent/compare/v1.0.1...v1.0.4) (2022-12-31)

### [1.0.3](https://github.com/AlOaks/minter-agent/compare/v1.0.1...v1.0.3) (2022-12-31)

### [1.0.2](https://github.com/AlOaks/minter-agent/compare/v1.0.1...v1.0.2) (2022-12-31)

### 1.0.2 (2022-12-31)


src/types.ts
import { Amount, NFTOffer } from "xrpl/dist/npm/models/common/index";

export type NFTCategoryType =
  | "art"
  | "motion"
  | "music"
  | "metaverse"
  | "sports"
  | "others"
  | "tradingcards"
  | "collectibles";

export type Mode = "devnet" | "mainnet" | "testnet";

export interface AcceptOfferOptions {
  isBuy: boolean;
}

export interface ParsedAmount {
  value: string | number;
  currency: string;
  issuer?: string;
}

export interface Bid {
  expiration?: number | Date | string;
  amount: ParsedAmount;
}
export interface NFTSaleOptions {
  amount: Amount;
  expiration?: number | Date | string;
  destination?: string;
}

export interface Price {
  [key: string]: number | string;
}

export interface Stats {
  "7d": number | string;
  "24h": number | string;
  "30d": number | string;
  all_time: number | string;
}

export interface StatsInterface {
  [key: string]: Stats;
}

export type NFTActionType =
  | "nft_minted"
  | "nft_offer_created"
  | "nft_sale_started"
  | "nft_sale_cancelled"
  | "nft_offer_cancelled"
  | "nft_sold"
  | "nft_transferred"
  | "nft_destroyed";

export interface NFTAction {
  type: NFTActionType;
  account: string;
  id: string;
  nft_id: string;
  txid: string;
  happened_at: string;
  standard: "XLS-20d" | "XLS-14d";
  from_account?: string;
  to_account?: string;
  quote_amount?: number | string;
  quote_currency?: string;
  quote_issuer?: string;
  is_standard_migration?: boolean;
  minter?: string;
}

export interface CollectionTradingData {
  floor_price: Price;
  nft_count: number;
  owner_count: number;
  volume: StatsInterface;
}

export interface NFT {
  Flags: number;
  Issuer: string;
  NFTokenID: string;
  NFTokenTaxon: number;
  URI?: string;
  nft_serial: number;
}

export interface BrokeredModeArgs {
  nft_id: string;
  sell_offer: NFTOffer;
  buy_offer: NFTOffer;
  max_broker_fee?: boolean;
  broker_fee?: Amount;
}

export interface NFTOffers {
  buy_offers: NFTOffer[];
  sell_offers: NFTOffer[];
}

export interface NFTActionsOptions {
  before_id?: number | string;
  limit?: number;
  types?: NFTActionType[];
}

export interface NFTMetadata {
  animation_url?: string;
  category: NFTCategoryType;
  content_type: string;
  description: string;
  external_url: string;
  image_url: string;
  is_explicit: boolean;
  md5hash: string;
  name: string;
}

export interface NFTData {
  id: string;
  standard: string;
  collection_id: string;
  minter: string;
  owner: string;
  ipfs_cid: string;
  md5_hash: string;
  minted_txid: string;
  metadata: NFTMetadata;
}

export interface NFTClio {
  nft_id: string;
  ledger_index: number;
  owner: string;
  issuer: string;
  is_burned?: boolean;
  flags?: number;
  transfer_fee?: number;
  nft_taxon?: number;
  nft_sequence?: number;
  uri?: string;
  validated: boolean;
}

export interface FullNFTData {
  sologenic_info: NFTData | null;
  xrpl_info: NFTClio;
}

export interface SignTransactionOptions {
  autofill?: boolean;
  withClient?: boolean;
}

export interface MintMultipleCopiesOptions {
  numberOfCopies: number;
  autoBurn?: boolean;
  onBehalf?: string;
}

export interface MintOptions {
  autoBurn?: boolean;
  onBehalf?: string;
}

export interface MintMultipleCopiesResult {
  copies_minted: number;
  nfts: NFTokenMintResult[];
  error?: any;
}

export interface NFTokenMintResult {
  mint_tx_hash: string;
  NFTokenID: string;
}

export interface CollectionData {
  name: string;
  cover?: Buffer;
  thumbnail?: Buffer;
  description?: string;
  transfer_fee?: number;
}

export interface BurnConfiguration {
  burn_amount: string | number;
  burn_amount_issuance?: string | number;
  burn_amount_market_index?: string | number;
  burn_currency: string;
  burn_issuer: string;
}

export interface BurnResult {
  address: string;
  hash: string;
  type?: string;
  burns_count: number;
  validated?: boolean;
}

export interface NFTSlot {
  uid: string;
  [key: string]: unknown;
}

export interface NFTAttribute {
  trait_type: string;
  value: string | number;
  max_value?: string | number;
}

export interface NFTPayload {
  file: Buffer;
  thumbnail: Buffer;
  name: string;
  category: NFTCategoryType;
  is_explicit: boolean;
  only_xrp: boolean;
  transfer_fee?: number; // A number between 0 and 50000 i.e to get 10% royalty transfer_fee must be 10000
  description?: string;
  external_url?: string;
  attributes?: NFTAttribute[];
}

export interface SologenicNFTManagerProps extends SologenicBaseProps {}

export interface SologenicNFTTraderProps extends SologenicBaseProps {}

export interface SologenicBaseProps {
  mode: Mode;
  xrpl_node: string;
}

export interface Collection {
  uid: string;
  issuer: string;
  nfts: NFTSlot[];
  [key: string]: unknown;
}
